### PR TITLE
soundwire: generic_bandwidth_allocation: optimize bandwidth for multiple Slaves on single link

### PR DESCRIPTION
**Pull request by RanderWang**
Set the same block offset to multiple Slaves on single link when these
slaves are working for one pcm stream. For playback, provide the same
data by the same block offset to both slaves, and let them extract what
they need. The codec might get sensor information to figure out how
to recombined the channels depending on the device position/orientation.

The first PR is at https://github.com/thesofproject/linux/pull/1712. Now let's wake up the thread for customer request.

### CODE REVIEW

File: drivers/soundwire/generic_bandwidth_allocation.c [hunk 1]
```diff
@@ -46,6 +46,9 @@ static void sdw_compute_slave_ports(struct sdw_master_runtime *m_rt,
 	port_bo = t_data->block_offset;
 
 	list_for_each_entry(s_rt, &m_rt->slave_rt_list, m_rt_node) {
+		if (m_rt->direction == SDW_DATA_DIR_TX)
+			port_bo = t_data->block_offset;
```
Comment (bardliao) [inline drivers/soundwire/generic_bandwidth_allocation.c]: This seems to be only valid when all slaves on the link belong to the same group. Can we also check slave group id and use the same prot_bo for the slaves with the same group?

Comment (plbossart) [reply drivers/soundwire/generic_bandwidth_allocation.c]: @bardliao the stream is handled by a Dai-link, so if we have 2 slaves on the same link, they will be part of the same stream, so the code above is safe by construction. If we have 2 slaves on the same link that are completely independent, they would not be part of the same stream.

**Comment by plbossart** [comment]
@RanderWang can I suggest a slightly different solution (untested, proof of concept only)

````diff
diff --git a/drivers/soundwire/generic_bandwidth_allocation.c b/drivers/soundwire/generic_bandwidth_allocation.c
index 7a747a302caf..57efdf40f293 100644
--- a/drivers/soundwire/generic_bandwidth_allocation.c
+++ b/drivers/soundwire/generic_bandwidth_allocation.c
@@ -42,6 +42,7 @@ static void sdw_compute_slave_ports(struct sdw_master_runtime *m_rt,
        struct sdw_port_runtime *p_rt;
        int port_bo, sample_int;
        unsigned int rate, bps, ch = 0;
+       unsigned int slave_total_ch;
 
        port_bo = t_data->block_offset;
 
@@ -50,6 +51,8 @@ static void sdw_compute_slave_ports(struct sdw_master_runtime *m_rt,
                bps = m_rt->stream->params.bps;
                sample_int = (m_rt->bus->params.curr_dr_freq / rate);
 
+               slave_total_ch = 0;
+
                list_for_each_entry(p_rt, &s_rt->port_list, port_node) {
                        ch = sdw_ch_mask_to_ch(p_rt->ch_mask);
 
@@ -67,6 +70,18 @@ static void sdw_compute_slave_ports(struct sdw_master_runtime *m_rt,
                                             SDW_PORT_DATA_MODE_NORMAL);
 
                        port_bo += bps * ch;
+                       slave_total_ch += ch;
+               }
+
+               if ((m_rt->direction == SDW_DATA_DIR_TX) &&
+                   (m_rt->stream->params.ch_count == slave_total_ch)) {
+                       /*
+                        * Slave devices were configured to access all channels
+                        * of the stream, which indicates that they operate in
+                        * 'mirror mode'. Make sure we reset the port offset for the
+                        * next device in the list
+                        */
+                       port_bo = t_data->block_offset;
                }
        }
 }
````

This would help preserve the current behavior if the Slave driver is configured with different ch_masks.

[pr1830.diff.txt](https://github.com/thesofproject/linux/files/4277431/pr1830.diff.txt)

File: drivers/soundwire/generic_bandwidth_allocation.c (hunk 1 above)
Comment (bardliao) [reply drivers/soundwire/generic_bandwidth_allocation.c]: Thanks @plbossart for the explanation

**Comment by RanderWang** [comment]
> @RanderWang can I suggest a slightly different solution (untested, proof of concept only)
> 
> ```diff
> diff --git a/drivers/soundwire/generic_bandwidth_allocation.c b/drivers/soundwire/generic_bandwidth_allocation.c
> index 7a747a302caf..57efdf40f293 100644
> --- a/drivers/soundwire/generic_bandwidth_allocation.c
> +++ b/drivers/soundwire/generic_bandwidth_allocation.c
> @@ -42,6 +42,7 @@ static void sdw_compute_slave_ports(struct sdw_master_runtime *m_rt,
>         struct sdw_port_runtime *p_rt;
>         int port_bo, sample_int;
>         unsigned int rate, bps, ch = 0;
> +       unsigned int slave_total_ch;
>  
>         port_bo = t_data->block_offset;
>  
> @@ -50,6 +51,8 @@ static void sdw_compute_slave_ports(struct sdw_master_runtime *m_rt,
>                 bps = m_rt->stream->params.bps;
>                 sample_int = (m_rt->bus->params.curr_dr_freq / rate);
>  
> +               slave_total_ch = 0;
> +
>                 list_for_each_entry(p_rt, &s_rt->port_list, port_node) {
>                         ch = sdw_ch_mask_to_ch(p_rt->ch_mask);
>  
> @@ -67,6 +70,18 @@ static void sdw_compute_slave_ports(struct sdw_master_runtime *m_rt,
>                                              SDW_PORT_DATA_MODE_NORMAL);
>  
>                         port_bo += bps * ch;
> +                       slave_total_ch += ch;
I print the value in driver:     ch = 2

````
> +               }
> +
> +               if ((m_rt->direction == SDW_DATA_DIR_TX) &&
> +                   (m_rt->stream->params.ch_count == slave_total_ch)) {
````
m_rt->stream->params.ch_count = 4, slave_total_ch = 2 (each slave have only one port enabled), so this judgement is false. 
ch_count is modified at https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/stream.c#L1176
````
> +                       /*
> +                        * Slave devices were configured to access all channels
> +                        * of the stream, which indicates that they operate in
> +                        * 'mirror mode'. Make sure we reset the port offset for the
> +                        * next device in the list
> +                        */
> +                       port_bo = t_data->block_offset;
>                 }
>         }
>  }
> ```
> 
> This would help preserve the current behavior if the Slave driver is configured with different ch_masks.
> 
> [pr1830.diff.txt](https://github.com/thesofproject/linux/files/4277431/pr1830.diff.txt)

**Comment by plbossart** [comment]
> m_rt->stream->params.ch_count = 4, slave_total_ch = 2 (each slave have only one port enabled), so this judgement is false.

That's good feedback, but allow me to bring a counter argument:

In the case with 2 devices on the same link, we will only use one CPU dai and the stream should have only 2 channels. There's no demux in the firmware.

If the stream is shown as using 4 channels then something is quite wrong. we would allocate bandwidth for something that's not needed. Can you help clarify how we end-up with a 4-ch stream? What is the cpu-dai/codec-dai mapping?

**Comment by RanderWang** [comment]
> This might be perfectly ok in soundwire tree, but to me the commit is hard to follow:
> "Set the same dp offset to multiple Slaves"
> 
> ... what is "dp offset"? I see the code change "port_po" and use "t_data->block_offset", but how that relates to "dp offset". I checked log history of drivers/soundwire/ and saw no commit messages discussing "dp offset". In case this is common term via soundwire specs, then this is ok and feel free to ignore the comment.

hi Kai, "t_data->block_offset" are finally set to dp_offsetCtrl register in cadance IP.  I focused on HW, so I indulged with dp offset. According to sdw spec, this should be block offset.

**Comment by RanderWang** [comment]
> > m_rt->stream->params.ch_count = 4, slave_total_ch = 2 (each slave have only one port enabled), so this judgement is false.
> 
> That's good feedback, but allow me to bring a counter argument:
> 
> In the case with 2 devices on the same link, we will only use one CPU dai and the stream should have only 2 channels. There's no demux in the firmware.
> 
> If the stream is shown as using 4 channels then something is quite wrong. we would allocate bandwidth for something that's not needed. Can you help clarify how we end-up with a 4-ch stream? What is the cpu-dai/codec-dai mapping?

Now ->stream->params.ch_count is used when there are two or more port enabled in master, so now it doesn't affect this case, which includes two 1308s with only one port.   Now there is one cpu dai (SDW1 PIN2) and two codec dais,  1308-L, 1308-R.

please check: https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/generic_bandwidth_allocation.c#L96-L122

**Comment by RanderWang** [comment]
Now we calculate bandwidth at https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/stream.c#L1508-L1512.  m_rt->ch_count is used, not stream->params.ch_count  and m_rt->ch_count is set by stream_config->ch_count, which is set to 2 by codec driver.

**Comment by plbossart** [comment]
> Now we calculate bandwidth at https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/stream.c#L1508-L1512. m_rt->ch_count is used, not stream->params.ch_count and m_rt->ch_count is set by stream_config->ch_count, which is set to 2 by codec driver.

can you clarify how we end-up with 4 channels?

m_rt->ch_count should be 2 and so should stream->ch_count.

**Comment by RanderWang** [comment]
> > Now we calculate bandwidth at https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/stream.c#L1508-L1512. m_rt->ch_count is used, not stream->params.ch_count and m_rt->ch_count is set by stream_config->ch_count, which is set to 2 by codec driver.
> 
> can you clarify how we end-up with 4 channels?
> 
> m_rt->ch_count should be 2 and so should stream->ch_count.

@plbossart ,  In my comments before: ch_count is modified at https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/stream.c#L1176
since there are two 1308,  stream will be added twice, so it was added to 4. The idea is FW should do demux.

**Comment by RanderWang** [comment]
@plbossart according to stream.c,  it seems that the author assumed: each slave only work on one channel. please check my updated PR

**Comment by RanderWang** [comment]
> I'm waiting for @plbossart and @RanderWang to close the open discussion, otherwise the patch looks ok to me now.
> 
> Minor nits in the commit message:
> 
> * the summary could be more to the point -- if I understood correctly, this patch does not touch the actual bandwidth allocation logic, but instead only fixes setting the "block offset" in the scenario of multi streams on a single link -- the summary leads me to believe you are changing bandwidth calculation somehow but that is not the case which is misleading
> * spelling: recombined -> recombine
> * spelling: don't start new-lines with a "," comma

@kv2019i  thanks for your comments. I will update the comments when we get a conclusion.

**Comment by RanderWang** [comment]
@plbossart  @kv2019i  update my patch.
